### DISCUSSION TITLE
Read current Armoury Crate control mode?

### DISCUSSION THREAD

**Discussion by eried** [seerge/g-helper#2981] [Q&A] [answered]
Hi, this is not a bug or suggestion, but related with the code, I am trying to detect which is the current mode the USB controller is on the ROG Ally portable devices

**There is something that g-helper and armory crate leave permanently while switching that can be detected?**

Something like the reverse of:

https://github.com/seerge/g-helper/blob/50f0c82b701ae350eb5959928e24b488be25b322/app/Ally/AllyControl.cs#L652

**Comment by seerge** [comment]
@eried hello,

1. Controller mode is not "permanent", device resets all it settings on reboot :)
2. There are no way to "read" current mode, only set it. This is how vast majority of settings work on asus devices.

**Reply by eried** [reply]
thanks for the insight!

so if armoury crate=`auto` and g-helper=`controller`, when switching to a window that armoury crate applies `desktop` there is no way to g-helper to ever know it was switched away from `controller` mode, right?

**Reply by eried** [reply]
I am just trying to make a launcher for emulatorstation, I hate to switch mode manually, but since emulatorstation spawns emulators, armorycrate interferes switching mode 😅

**Reply by seerge** [reply, marked as answer]
There is no such hardware mode as `Auto`.  

G-Helper monitors FPS metric from AMD driver and sets `Gamepad` mode for controller when it sees FPS > 0 (3 times in a row) and sets back `Mouse` when FPS counter reports NULL or 0.

https://github.com/seerge/g-helper/blob/50f0c82b701ae350eb5959928e24b488be25b322/app/Ally/AllyControl.cs#L378

AC's implementation of `Auto` is different, as I understand it's based on it's own game-launcher (so it just knows that game is running), but it does similar thing.

---

There is no way to read current controller setting from firmware.
